File: src/commands/env/pull.rs
use colored::*;
use std::env;
use std::fs;
use std::process::Command;

use crate::commands::setup::utils as setup_utils;
use crate::vault;

use super::utils as env_utils;

pub fn run() {
    if let Err(e) = run_inner() {
        eprintln!("{} {}", "Pull failed:".red().bold(), e);
        crate::terminal::cleanup_and_exit(1);
    }
}

fn run_inner() -> Result<(), String> {
    let cwd = env::current_dir()
        .map_err(|e| format!("Cannot read current directory: {}", e))?;

    // Derive project ID
    let (project_id, used_fallback) = env_utils::derive_project_id(&cwd)?;
    if used_fallback {
        println!(
            "  {} {}",
            "!".yellow().bold(),
            format!(
                "No git remote found. Using folder name '{}' as project identifier.",
                project_id
            ).dimmed()
        );
    }
    println!(
        "  {} {}",
        "Project:".dimmed(),
        project_id.cyan().bold()
    );

    // Resolve vault and pull latest
    let vault_dir = vault::vault_dir();
    let vault_dir_str = vault_dir.display().to_string();

    let sp = setup_utils::spinner("Pulling latest from vault...");
    let _ = setup_utils::run_git(&vault_dir_str, &["pull"]);
    sp.finish_and_clear();

    // Check if project exists in vault
    let src_dir = env_utils::envs_dir(&vault_dir, &project_id);
    if !src_dir.exists() {
        return Err(format!(
            "No .env files stored for project '{}'.\n  \
             Push first with: keypick env push",
            project_id
        ));
    }

    // Find encrypted .env files in vault
    let entries: Vec<_> = fs::read_dir(&src_dir)
        .map_err(|e| format!("Failed to read {}: {}", src_dir.display(), e))?
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.file_name()
                .to_string_lossy()
                .starts_with(".env")
        })
        .collect();

    if entries.is_empty() {
        return Err(format!(
            "No .env files found in vault for project '{}'.",
            project_id
        ));
    }

    // Decrypt each file and write to CWD
    let sp = setup_utils::spinner("Decrypting .env files...");
    let mut written = Vec::new();
    let mut errors = Vec::new();

    for entry in &entries {
        let file_name = entry.file_name();
        let file_name_str = file_name.to_string_lossy();
        let src_path = entry.path();
        let dest_path = cwd.join(file_name_str.as_ref());

        let existed = dest_path.exists();

        let output = Command::new("sops")
            .args([
                "--decrypt",
                "--input-type", "dotenv",
                "--output-type", "dotenv",
            ])
            .arg(&src_path)
            .output()
            .map_err(|e| format!("Failed to run sops: {}", e))?;

        if output.status.success() {
            fs::write(&dest_path, &output.stdout)
                .map_err(|e| format!("Failed to write {}: {}", dest_path.display(), e))?;
            written.push((file_name_str.to_string(), existed));
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            errors.push(format!("{}: {}", file_name_str, stderr.trim()));
        }
    }

    sp.finish_and_clear();

    // Report results
    for (name, existed) in &written {
        let action = if *existed { "overwritten" } else { "created" };
        println!(
            "  {} {} ({})",
            "✓".green().bold(),
            name.cyan(),
            action.dimmed()
        );
    }

    for err in &errors {
        println!("  {} {}", "✗".red(), err);
    }

    if written.is_empty() {
        return Err("All files failed to decrypt.".to_string());
    }

    println!(
        "\n  {} {} .env file(s) pulled for {}",
        "✓".green().bold(),
        written.len().to_string().cyan().bold(),
        project_id.cyan()
    );

    Ok(())
}

File: ts/src/commands/env/pull.ts
// `keypick env pull` — decrypt .env files from the vault to the current project.

import { existsSync, readdirSync, statSync, writeFileSync } from "node:fs";
import path from "node:path";
import process from "node:process";
import { spawnSync } from "node:child_process";
import chalk from "chalk";
import * as vault from "../../lib/vault.ts";
import * as terminal from "../../lib/terminal.ts";
import * as setupUtils from "../setup/utils.ts";
import * as envUtils from "./utils.ts";

export async function run(): Promise<void> {
  try {
    await runInner();
  } catch (e) {
    console.error(chalk.red.bold("Pull failed:"), (e as Error).message);
    terminal.cleanupAndExit(1);
  }
}

async function runInner(): Promise<void> {
  const cwd = process.cwd();
  const { id: projectId, usedFallback } = envUtils.deriveProjectId(cwd);
  if (usedFallback) {
    console.log(
      `  ${chalk.yellow.bold("!")} ${chalk.dim(
        `No git remote found. Using folder name '${projectId}' as project identifier.`,
      )}`,
    );
  }
  console.log(`  ${chalk.dim("Project:")} ${chalk.cyan.bold(projectId)}`);

  const vaultDir = await vault.vaultDir();

  const sp = setupUtils.spinner("Pulling latest from vault...");
  try {
    setupUtils.runGit(vaultDir, ["pull"]);
  } catch {
    // best-effort
  }
  sp.stop();

  const srcDir = envUtils.envsDir(vaultDir, projectId);
  if (!existsSync(srcDir)) {
    throw new Error(
      `No .env files stored for project '${projectId}'.\n  Push first with: keypick env push`,
    );
  }

  const entries = readdirSync(srcDir)
    .map((name) => ({ name, full: path.join(srcDir, name) }))
    .filter((e) => {
      try {
        return statSync(e.full).isFile() && e.name.startsWith(".env");
      } catch {
        return false;
      }
    });

  if (entries.length === 0) {
    throw new Error(`No .env files found in vault for project '${projectId}'.`);
  }

  const sp2 = setupUtils.spinner("Decrypting .env files...");
  const written: { name: string; existed: boolean }[] = [];
  const errors: string[] = [];

  for (const entry of entries) {
    const destPath = path.join(cwd, entry.name);
    const existed = existsSync(destPath);
    const result = spawnSync(
      "sops",
      ["--decrypt", "--input-type", "dotenv", "--output-type", "dotenv", entry.full],
      { encoding: "buffer" },
    );
    if (result.error) {
      sp2.stop();
      throw new Error(`Failed to run sops: ${result.error.message}`);
    }
    if (result.status === 0) {
      try {
        writeFileSync(destPath, result.stdout!);
        written.push({ name: entry.name, existed });
      } catch (e) {
        sp2.stop();
        throw new Error(`Failed to write ${destPath}: ${(e as Error).message}`);
      }
    } else {
      const stderr = (result.stderr?.toString("utf8") ?? "").trim();
      errors.push(`${entry.name}: ${stderr}`);
    }
  }
  sp2.stop();

  for (const { name, existed } of written) {
    const action = existed ? "overwritten" : "created";
    console.log(`  ${chalk.green.bold("✓")} ${chalk.cyan(name)} (${chalk.dim(action)})`);
  }
  for (const err of errors) {
    console.log(`  ${chalk.red("✗")} ${err}`);
  }

  if (written.length === 0) {
    throw new Error("All files failed to decrypt.");
  }

  console.log(
    `\n  ${chalk.green.bold("✓")} ${chalk.cyan.bold(written.length)} .env file(s) pulled for ${chalk.cyan(projectId)}`,
  );
}
